crates/bevy_ecs/src/schedule/run_criteria.rs

```diff
@@ -0,0 +1,360 @@
+    }
+
+    fn after(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        new_run_criteria_descriptor(self).after(label)
+    }
+}
+
+impl<S> RunCriteriaDescriptorCoercion for S
+where
+    S: System<In = (), Out = ShouldRun>,
+{
+    fn label(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        new_run_criteria_descriptor(Box::new(self)).label(label)
+    }
+
+    fn before(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        new_run_criteria_descriptor(Box::new(self)).before(label)
+    }
+
+    fn after(self, label: impl RunCriteriaLabel) -> RunCriteriaDescriptor {
+        new_run_criteria_descriptor(Box::new(self)).after(label)
+    }
+}
+
+pub struct RunCriteria {
+    label: BoxedRunCriteriaLabel,
+}
+
+impl RunCriteria {
+    /// Constructs a new run criteria that will retrieve the result of the criteria `label`
+    /// and pipe it as input to `system`.
+    pub fn pipe(
+        label: impl RunCriteriaLabel,
+        system: impl System<In = ShouldRun, Out = ShouldRun>,
+    ) -> RunCriteriaDescriptor {
+        label.pipe(system)
+    }
+}
+
+pub trait RunCriteriaPiping {
+    fn pipe(self, system: impl System<In = ShouldRun, Out = ShouldRun>) -> RunCriteriaDescriptor;
+}
+
+impl RunCriteriaPiping for BoxedRunCriteriaLabel {
+    fn pipe(self, system: impl System<In = ShouldRun, Out = ShouldRun>) -> RunCriteriaDescriptor {
+        RunCriteriaDescriptor {
+            system: RunCriteriaSystem::Chained(Box::new(system)),
+            label: None,
+            before: vec![],
+            after: vec![self],
+        }
+    }
+}
+
+impl<L> RunCriteriaPiping for L
+where
+    L: RunCriteriaLabel,
+{
+    fn pipe(self, system: impl System<In = ShouldRun, Out = ShouldRun>) -> RunCriteriaDescriptor {
+        RunCriteriaDescriptor {
+            system: RunCriteriaSystem::Chained(Box::new(system)),
+            label: None,
+            before: vec![],
+            after: vec![Box::new(self)],
+        }
+    }
+}
+
+pub struct RunOnce {
+    ran: bool,
+    system_id: SystemId,
+    archetype_component_access: Access<ArchetypeComponentId>,
+    component_access: Access<ComponentId>,
+}
+
+impl Default for RunOnce {
+    fn default() -> Self {
+        Self {
+            ran: false,
+            system_id: SystemId::new(),
+            archetype_component_access: Default::default(),
+            component_access: Default::default(),
+        }
+    }
+}
+
+impl System for RunOnce {
+    type In = ();
+    type Out = ShouldRun;
+
+    fn name(&self) -> Cow<'static, str> {
+        Cow::Borrowed(std::any::type_name::<RunOnce>())
+    }
+
+    fn id(&self) -> SystemId {
+        self.system_id
+    }
+
+    fn new_archetype(&mut self, _archetype: &Archetype) {}
+
+    fn component_access(&self) -> &Access<ComponentId> {
+        &self.component_access
+    }
+
+    fn archetype_component_access(&self) -> &Access<ArchetypeComponentId> {
+        &self.archetype_component_access
```

Comment:
If we added a const `Access::new` constructor this could be `&Access::new()`.
This should be possible since https://github.com/bevyengine/bevy/commit/42924d22272fb0a38a01bab1116ad7b6f8865e8a as `fixedbitset` now has a `const FixedBitSet::new`.

Reply:
This PR simply moves this code into a separate module. While that could be a good idea, it's out of scope here.